examples/microkit_sddf_net/apps/debugger/debugger.c

```diff
@@ -79,6 +82,71 @@ static bool debugger_initialized = false;
 
 #define NUM_DEBUGEES 2
 
+// Define table data regions for our child
+microkit_uint64_t table_metadata[64];
+microkit_uint64_t table_data[0x800 * (NUM_DEBUGEES + 100)];
+
+microkit_uint64_t walk_table(microkit_uint64_t *start, uintptr_t addr, int lvl)
+{
+    // Get the 9 bit index for this level.
+    uintptr_t index = 0;
+    if (lvl == 1) {
+        // Index into the PUD
+        index = (addr & ((uintptr_t)0x1ff << 30)) >> 30;
+    } else if (lvl == 2) {
+        // Index into the PD
+        index = (addr & ((uintptr_t)0x1ff << 21)) >> 21;
+    } else if (lvl == 3) {
+        // Index into the PT
+        index = (addr & ((uintptr_t)0x1ff << 12)) >> 12;
+    }
+    if (lvl == 3) {
+       return start[index];
+    } else {
+        walk_table(&table_data[start[index]/8], addr, lvl + 1);
```

Comment:
Instead of making it recursive, just use a switch with fall-through instead of if-else, and update start along the way? But if it always starts with lvl 0 it can be even simpler. It would make more sense if lvl was the level you are looking for, than starting level.